File: Sorting/Bubble Sort/Kotlin/bubble_sort.kt
package com.freaklius.kotlin.algorithms.sort

/**
 * Bubble sort algorithm
 * AveragePerformance = O(n^2)
 */
public class BubbleSort : SortAlgorithm{

    public override fun sort(arr: Array<Long>): Array<Long> {
        var swappedElements : Boolean;
        do {
            swappedElements = false;
            for (i in 0..arr.size - 2){
                if (arr[i] > arr[i + 1]){
                    swap(arr, i, i + 1);
                    swappedElements = true;
                }
            }
        } while (swappedElements);
        return arr;
    }

    public override fun getName(): String {
        return "BubbleSort";
    }
}

File: Sorting/Bubble Sort/Dart/bubble_sort.dart
void bubbleSort(List list) 
{
  if (list == null || list.length == 0) 
    return;

  int num = list.length;
  int i, step;
  for (step = 0; step < num; step++) {
    for ( i = 0; i < num - step - 1; i++) {
      if (list[i] > list[i + 1]) {
        swap(list, i);
      }
    }
  }
}

void swap(List list, int i) {
  int temp = list[i];
  list[i] = list[i+1];
  list[i+1] = temp;
}